Exclude banks from Total scattering merge banks

sf1919 opened the pull request:
**Description of work.**
Users are now able to exclude specific banks from the merge banks option in total scattering by setting x-min and x-max for that bank to -1. X is positive so this is a good way of denoting they need to be excluded.

This PR adds this functionality, including a check that the exclusion is in the same place on x-min and x-max. Documentation has been added (including an example) to show how this functionality works. Tests have been added too.

**To test:**
1. Unzip this zip file somewhere:
[POLARISPowderDiffractionFolderSmall.zip](https://github.com/mantidproject/mantid/files/9103459/POLARISPowderDiffractionFolderSmall.zip)

2. Update the paths in the file polaris_config_example.yaml so they match where you unzipped the file and also update config_file_path in the script below

3. Run the following script in workbench and check it completes.
```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

from isis_powder import polaris, SampleDetails

config_file_path = r"C:\mantid-development\POLARISPowderDiffraction\polaris_config_example.yaml"
polaris = polaris.Polaris(config_file=config_file_path, user_name="test", mode="PDF")

sample_details = SampleDetails(height=4.0, radius=0.2985, center=[0, 0, 0], shape='cylinder')
sample_details.set_material(chemical_formula='Si')
polaris.set_sample_details(sample=sample_details)

polaris.create_vanadium(first_cycle_run_no="98532", multiple_scattering=False)

polaris.focus(run_number="98533", input_mode='Summed')

q_lim = [[2.5, 3, 4, 6, 7], [3.5, 5, 7, 11, 40]]

ws = mtd['98533-ResultTOF_2']
ws *= 100

polaris.create_total_scattering_pdf(run_number="98533", merge_banks=True,q_lims=q_lim, 
                                    freq_params=[1.4], pdf_type="g(r)")
```

4. Rename the pdf workspace so that it won't be overwritten when you re-run the code above.
5. Change the qlim line in the code above to the line below and then re-run it
```q_lim = [[2.5, -1, 4, 6, 7], [3.5, -1, 7, 11, 40]]```
<!-- Instructions for testing. -->

6. Plot the spectrum in both of the pdf workspaces (the new one and the one renamed in step 4). They should be different.


Fixes #33830 


---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

sf1919 committed:
Add exclude_banks method to MatchAndMergeWorkspaces.py

sf1919 committed:
Added documentation for #33830

sf1919 committed:
Raise error if -1 put in incorrect place

- If user fails to put -1 to exclude a bank in the same place on both lists
(e.g. first in position 1 and the second in position 2 of their respective lists),
an error will be raised.
- Test added for this scenario

sf1919 committed:
Added test for excluding banks when merging

DannyHindson commented on docs/source/techniques/ISISPowder-Polaris-v1.rst:
```diff
@@ -190,8 +190,8 @@ The output PDF can be customized with the following parameters:
 - By calling with `merge_banks=True` a PDF will be generated based on the weighted sum of the detector
   banks performed using supplied Q limits `q_lims=q_limits`, q_limits can be in the form of an array or
   with shape (2, x) where x is the number of detectors, or a string containing the directory
```
I think this should say where x is the number of banks rather than detectors? Not your piece of text but noticed it because you've repeated "detectors" in the extra text you've added. Please can you change both to banks

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/MatchAndMergeWorkspaces.py:
```diff
@@ -145,6 +147,21 @@ def fit_x_lims_to_match_histogram_bins(ws_conjoined, x_min, x_max):
+        index_to_remove = []
+        for i in range(x_min.size):
+            if x_min[i] == -1:
+                if x_max[i] == -1:
```
This approach of checking x_min before checking x_max doesn't catch a situation where there's an extra -1 in the x_max list. I think you probably need to check `if x_min[i] == -1 and x_max[i]==-1` and then in the else you can check if either is true and raise the error. I happened to do a test with these qlims when I spotted this:
```
q_lim = [[2.5, -1, 4, 6, 7], [3.5, -1, -1, 11, 40]]
```

DannyHindson reviewed (changes requested):
Just a couple of small things but the main change works fine

sf1919 committed:
Changes after review

- Updated documentation
- Revised check for -1 in same place in both lists and updated the associated test

DannyHindson reviewed (changes requested):
Please can you add a check that there is at least one bank left - probably at the end of exclude_banks? At the moment if you put -1 in all the slots in qlims it gives this error:

![image](https://user-images.githubusercontent.com/56295817/181510608-f3bab7d9-5b72-4eb7-b211-00152e873d7c.png)

Sorry - could perhaps have thought of that on my initial review. Last set of changes are all good thanks

sf1919 committed:
Further changes

- ensure that user provides information for at least one bank for merging to take place

DannyHindson reviewed (approved):
Another one done! Thanks for the changes and for adding the test as well
